### PR TITLE
JAMES-2733 [PROPOSAL] Optimize RabbitMQ MailQueue getSize

### PR DESCRIPTION
**Pull request by chibenwa**
Idea: Use a Per-Bucket mail counter projection

Upon enqueue:

 -  Increment the counter of the bucket the mail is assigned to

Upon dequeue:

 -  Decrement the counter of the bucket the mail is assigned to

Upon get size:

 -  Starting from the browseStart point...
 -  Sum all the bucket counts starting from browseStart

Another projection must be use to retrieve, given a mailKey, the associated bucket+slice.

Work remaining: Write unit tests for DAOs

### CODE REVIEW

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueBrowser.java [hunk 1]
```diff
@@ -111,6 +115,18 @@ public boolean hasNext() {
             .subscribeOn(Schedulers.parallel());
     }
 
+    Mono<Long> getSize(MailQueueName queueName) {
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueBrowser.java]: It's not related to queue browsing, so it should live elsewhere

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueMailDelete.java [hunk 2]
```diff
@@ -59,9 +63,15 @@
     Mono<Void> considerDeleted(MailKey mailKey, MailQueueName mailQueueName) {
         return deletedMailsDao
             .markAsDeleted(mailQueueName, mailKey)
+            .then(decrementBucketSize(mailKey, mailQueueName))
             .doOnNext(ignored -> maybeUpdateBrowseStart(mailQueueName));
     }
 
+    private Mono<Void> decrementBucketSize(MailKey mailKey, MailQueueName mailQueueName) {
+        return mailKeyToBucketDAO.retrieveBucket(mailQueueName, mailKey)
```
Comment (mbaechler) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueMailDelete.java]: I'm not too worry about writes but adding a read-before-write to this code path looks scary

**Comment by chibenwa** [comment]
> could we just rely on slices ?

Clever: we don't need the additional partitionning brought in by the bucket. Which spares some scalability concern: the complexity is then only proportional to the  sliceCount.

If we can infer the enqueued date from the mail we can save the additional projection. Maybe adding the slice in the MailReferenceDTO?

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueMailDelete.java (hunk 2 above)
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueMailDelete.java]: Agree. 

If we can infer the enqueued date from the mail we can save the additional projection. Maybe adding the slice in the MailReferenceDTO - and carry it along in the RabbitMQXXXItem?

**Comment by chibenwa** [comment]
Well, let's give a bit of context...

> it is reasonable to query the queue holder for that

**MailQueue** is a component in James that needs to act as a WorkQueue (messages shared between subscriber) but with additional features. Think getSize, browse, delete, clear. Implementing this component with Cassandra is inneficient, and with RabbitMQ, the expected features are impossible to be implemented. We thus decided to implement a complex design combining both components.

As a direct consequence of it, **RabbitMQ is not the queue holder**. RabbitMQ holds some informations that *may* had been deleted. Relying solely on it leads to invalid behaviours.

> Can you re-explain the bucket/slice mechanism please?

Slicing appears when you want to order items by time - it is a common technic when working with **time series**. Slicing allows us to only iterate on the latest elements of the queues - and solves the tumbstone iteration problems inherant to queues.

One big concern in Cassandra is avoiding **wide rows**. A wide row is when too much items shares the same primary key. Here that means that all the writes were handled by a single Cassandra node at any given time (the one holding queueName + sliceId ). Buckets is a good artificial way of spreading things up in the cluster.

> Going back to the user story "As an Admin, I want to be sure that my deletions request worked"

Please comment **on the right issue**. I did the effort to handle this in a separate PR and consider that bugfix orthogonal to this 'optimization work'.

> What we really need is an entry point to check dropped requests/messages.

We already have it: `MailQueueView::isPresent` which is basically used by dequeueue + browse. Now, this is not enough to achieve our purposes given the performance issues you encountered.

> I'm not really a big fan on adding complexity to go faster

Neither do I. That's why we did not introduced projections for the queue size initially.

Now we gained insight, and got pretty convinced the current `getSize` needs to be reworked in order to play well with a 1M+ queue size.

Not only we know this optimization is needed, but we have time to think about and come up with the right design - this proposal is a good base IMO that needs to be refined in order to get the right approach - see @mbaechler constructive comments.

> About complexity

I agree that mailQueue is a complex component where not fully understanding the design will lead to bugs (cf broken deletes) and inneficient code. I agree things are easy to get wrong.

Maybe it is time to document this complex design within a README in the code to explain concepts, justify choices.

**Comment by blackheaven** [comment]
Thanks for your clarifications.

> Please comment on the right issue. I did the effort to handle this in a separate PR and consider that bugfix orthogonal to this 'optimization work'.

It is the right issue, I'm convinced that we are not dealing with the right semantic for that.

> We already have it: MailQueueView::isPresent which is basically used by dequeueue + browse. Now, this is not enough to achieve our purposes given the performance issues you encountered.

I mean a WebAdmin endpoint

Again, if you want to keep this semantic, having a table of messages to drop instead of a deleted messages table would be enough to do the maths.

**Comment by chibenwa** [comment]
> Again, if you want to keep this semantic, having a table of messages to drop instead of a deleted messages table would be enough to do the maths.

If you do count present messages, you do just need to do a `sum`.

If you do count deleted message, then I guess `size = rabbitMQQueueSize - deletedMessages`. The **big** care here is that we need to be **sure** that the **deletedMessages mentionned above are 1. marked as deleted 2. still enqueued in RabbitMQ. This proposal would need to rework MailQueueueView to differenciate dequeues from delete (currently a dequeued message is deleted from the view) 2. involve the RabbitMQ admin API which proved to be crappy. So yes the math are correct - but that don't means it is easier.

> I mean a WebAdmin endpoint

What does the admin have to do in that?

> It is the right issue, I'm convinced that we are not dealing with the right semantic for that.

Do you suggest we should come back to the design table?

What's your proposals?

**Comment by blackheaven** [comment]
> `size = rabbitMQQueueSize - deletedMessages`

This will not work since `deletedMessages` are all the deleted messages and not only the enqueued one.

> > I mean a WebAdmin endpoint
> 
> What does the admin have to do in that?

If (s)he want's to ensure that the requested removals have been registered, that's the only way to do.

> > It is the right issue, I'm convinced that we are not dealing with the right semantic for that.
> 
> Do you suggest we should come back to the design table?
> 
> What's your proposals?

Instead of have a `deletedMessages` have a `dropMessages` table which only contains the messages remaining (enqueued) to be dropped. Every-time a message have been processed, the message is acknowledged and the row is removed from `dropMessages`.

It has to major benefits:

 1. Having a descent-size table (instead of a growing table)
 2. Having a fast way to query size `size = RabbitMQ.queueSize - dropMessages.size`

**Comment by chibenwa** [comment]
> Having a fast way to query size size = RabbitMQ.queueSize - dropMessages.size

You forget tumbstone : that will be slow. And you would end up with the queue like design on Cassandra - that we had been triing hard to avoid.

What's the issue with ever growing table?

> If (s)he want's to ensure that the requested removals have been registered, that's the only way to do.

Why couldn't I wait the completion of the task I just created?

If I'm not convinced then I can go with a browse...

I'm not sure getting the ManageableMailQueue more DB like by adding per-record read is necessarily the good way to go. I'd be cautious with that...

**Comment by blackheaven** [comment]
> You forget tumbstone : that will be slow.

Not compared to the current implementation.

> And you would end up with the queue like design on Cassandra - that we had been trying hard to avoid so far - which yes, makes things harder.

Not really, a queue has a notion of order, here it's an arbitrary access.

> What's the issue with ever growing table?

Performances, costs...

> > If (s)he want's to ensure that the requested removals have been registered, that's the only way to do.
> 
> Why couldn't I wait the completion of the task I just created?

That was your argument.

> I'm not sure getting the ManageableMailQueue more DB like by adding per-record read is necessarily the good way to go. I'd be cautious with that...

That's not my proposal.

Instead of reading in an ever growing table, as we currently do, we will read in a small table.

**Comment by mbaechler** [comment]
>>  You forget tumbstone : that will be slow.

> Not compared to the current implementation.

We won't create tables we queue-like access pattern We already know it will be a disaster.

> [about table size]

Table size is never an issue because you always address a line by a primary key, which is fast enough (probably 0(log(n)))

The point is we don't want to delete stuff in active slices (because of the tombstone problem) but there's no issue removing entire slices they are not in use anymore. So basically, we don't have a size problem at all.

**Comment by blackheaven** [comment]
> > > You forget tumbstone : that will be slow.
> 
> > Not compared to the current implementation.
> 
> We won't create tables we queue-like access pattern We already know it will be a disaster.

It's not the case at all since it's for rare cases.

> > [about table size]
> 
> Table size is never an issue because you always address a line by a primary key, which is fast enough (probably 0(log(n)))
> 
> The point is we don't want to delete stuff in active slices (because of the tombstone problem) but there's no issue removing entire slices they are not in use anymore. So basically, we don't have a size problem at all.

Is it currently implemented?

**Comment by chibenwa** [comment]
> It's not the case at all since it's for rare cases.

As far that I know getSize is called every 30s

**Comment by blackheaven** [comment]
> > It's not the case at all since it's for rare cases.
> 
> As far that I know getSize is called every 30s

I was talking about removing messages from queue.

In which context?

**Comment by chibenwa** [comment]
> In which context?

Metric collection

**Comment by blackheaven** [comment]
> > In which context?
> 
> Metric collection

That's catastrophic.

**Comment by blackheaven** [comment]
Anyway, after a talk with @mbaechler it seems that `deletedMessages` is used for `browse`.

Either we can add a field which be filtered on to get the messages remaining to be dropped, or create a `dropMessages` as a projection.

**Comment by chibenwa** [comment]
> create a dropMessages as a projection.

Again how do you know if the dropped message had been dequeued or not?

**Comment by blackheaven** [comment]
> > create a dropMessages as a projection.
> 
> Again how do you know if the dropped message had been dequeued or not? (akka if you should take it into account or ignore it)

Why would you need this information?

> > we can add a field which be filtered on to get the messages remaining to be dropped
> 
> How is it different from a browse - what exists today?

You not not go through all the deleteMessages just to get the size.

> > > > In which context?
> 
> > > Metric collection
> 
> > That's catastrophic.
> 
> Let me disagree - isn't it a valueable metric?
> 
> We can argue that having every node measuring it is a bad idea - but hey coordination, elections and the like are not easy as well...

At the cost of a high paid performance issue, I should have the choice of getting it or not.

**Comment by chibenwa** [comment]
> At the cost of a high paid performance issue, I should have the choice of getting it or not.

Looks like a constructive reply that can get us to a temporary workaround! Let's implement this now?

Concerning your proposed alternative algorithm, please come up with a written, step by step algorithm for all operations, as well as table structure. Until we got that, discussing it is IMO a loss of time.

**Comment by blackheaven** [comment]
Agreed

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueMailDelete.java (hunk 2 above)
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueMailDelete.java]: Let me answer you on this point:

MailQueueView storage informations is not transmitted over rabbitMQ for good reasons:

 - Publishing to rabbit is done before saving the mail in the view: this means the information is not available 'yet'. Reason consistency: if the rabbit call fails, and we reverse the order the mail will never be cleaned up from the view.

 - We don't wan't to leak Cassandra specific details in the Rabbit specific part of the code anyway - or at the price of costly abstractions.

Now thinking about this "cold-headed', this do not scare me much: it's only very simple metadata. After the queue (where being slow is OK BTW).

**Comment by chibenwa** [comment]
Will re-open a new version of this PR

**Pull request closed without merging**